libs/service-portal/access-control/delegations/src/hooks/useDomains.ts

```diff
@@ -0,0 +1,114 @@
+import { useEffect, useMemo, useState } from 'react'
+import { useLocale } from '@island.is/localization'
+import { useAuthDomainsQuery } from '@island.is/service-portal/graphql'
+import { ALL_DOMAINS, ISLAND_DOMAIN } from '../constants'
+import { useQueryParam } from '@island.is/service-portal/core'
+import { useLocation, useHistory } from 'react-router-dom'
+import { storageFactory } from '@island.is/shared/utils'
+
+const sessionStore = storageFactory(() => sessionStorage)
+
+export type DomainOption = {
+  label: string
+  value: string
+}
+
+/**
+ * This domain hook is used to fetch domains list for cuttent user as well as handle selection of the domain,
+ * either in query string or session storage.
+ *
+ * The priority is the following:
+ * 1. If there is a domain in query string and no session storage, use the query string
+ * 2. If there is a domain in query string and session storage, use the session storage.
+ * 3  If there is no domain in query string and session storage, use session storage.
+ * 4. If there is no domain in query string and no session storage, use the default domain, i.e. ISLAND_DOMAIN.
+ */
+export const useDomains = () => {
+  const { formatMessage, lang } = useLocale()
+  const location = useLocation()
+  const history = useHistory()
+  const displayNameQueryParam = useQueryParam('domain')
+  const [selectedDomainName, setSelectedDomainName] = useState<string | null>(
+    null,
+  )
+
+  const defaultLabel = formatMessage({
+    id: 'sp.access-control-delegations:all-domains',
+    defaultMessage: 'Öll kerfi',
+  })
+  const allDomainsOption = {
+    label: defaultLabel,
+    value: ALL_DOMAINS,
+  }
+
+  const { data, loading } = useAuthDomainsQuery({
```

Comment:
We received comment from Core team about graphql queries and not storing them in one central place:

*I know this is done almost everywhere, but we really shouldn't store queries in a central location and query everything. This takes the power away from GraphQL, and treats it like a Rest API.*

*Components should query the needed data. GraphQL will take care of merging those requests into a batch operation and remove redundant fields.*

Reply:
Good and valid point. 
There have been some debates in the past as I understood it on what would be the preferred way of doing the queries.
There are ups and downs to both ways that where discusses, i.e. let queries coexist inside component files or generate hooks in one or more places.

We evaluated that we would prefer typed queries and responses over having the queries where the components live.
In a perfect world we would like to run codegens on module bases so that we don't have all the hooks in one specific lib, but that dosen't scale well when generating types from the GraphQL API.